### PR TITLE
[std.traits] Add 'See also: __traits'

### PR DESCRIPTION
**Pull request by ntrel**
**Note: Originally this recommended using __traits instead, now it just says see also.**

---
That will reduce template instantiation overhead.
Note that some __traits accept type instances in addition to just types.

~~I've left `isNested` as that only applies to types not functions - ideally it would be named `isNestedType` to avoid confusion with `__traits(isNested)`.~~

@schveiguy

### CODE REVIEW

File: std/traits.d [hunk 1]
```diff
@@ -6161,7 +6163,7 @@ template BuiltinTypeOf(T)
 //::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::://
 
 /**
- * Detect whether `T` is a built-in boolean type.
+ * Detect whether `T` is a built-in boolean type or enum of boolean base type.
```
Comment (ntrel) [inline std/traits.d:6166]: Note: This was already de-facto documented in the unittest below.

File: std/traits.d [hunk 2]
```diff
@@ -3853,6 +3853,7 @@ package alias Identity(alias A) = A;
 /**
    Yields `true` if and only if `T` is an aggregate that defines
    a symbol called `name`.
+   $(NOTE Use `__traits(hasMember, T, name)` instead in new code.)
```
Comment (RazvanN7) [inline std/traits.d:3856]: Well, I find this situation a bit comical. Why do we even keep the abstraction if we recommend using `__traits`? From my perspective, we either deprecate the abstraction and justify this by the fact that you can use traits directly, or we leave it in the name of "__traits is ugly, users should use the stdlib abstractions".

**Comment by dkorpel** [comment]
I don't think there's much overhead in instantiating a template that's a thin wrapper over `__traits`. `__traits` has the double underscore prefix to make it uglier and less appealing to use directly in code, the opposite of what the new comment recommends.

**Comment by ntrel** [comment]
> I agree __traits shouldn't have the double underscore.

Well `traits` can't be a keyword as it's used in `std.traits`.

> there aren't enough savings here to justify recommending __traits usage in these cases

OK, changed to just 'see also'.

> is there a way we can detect in the compiler a pattern like alias x(args...) = __traits(something, args) and avoid the overhead completely? I know we avoid AliasSeq storage, can we utilize that same mechanism?

The AliasSeq detection slowed down the compiler a bit according to a forum reply at the time. Also you mentioned they can have different semantics.

**Comment by ntrel** [comment]
> The AliasSeq detection slowed down the compiler a bit

BTW we really should have a sequence literal in the language, relying on a template is not only inefficient but annoying to have to import it, and it would also make examples cleaner in the documentation when learning about sequences. A sequence is a fundamental building block in the language, it's not just a template parameter. I would like to add `__Seq(args)`.

**Review by schveiguy — APPROVED** [review]
Looks good.

A bit of bikeshedding:

1. Can the "see also" sections be links?
2. Should it say "see also" or "equivalent to"?

File: std/traits.d [hunk 3]
```diff
@@ -7623,14 +7637,15 @@ template isCallable(alias callable)
 
 
 /**
-Detect whether `T` is an abstract function.
+Detect whether `S` is an abstract function.
```
Comment (RazvanN7) [inline std/traits.d:7640]: This change seems unrelated.

**Review by RazvanN7 — APPROVED** [review]

**Comment by ntrel** [comment]
>     Can the "see also" sections be links?

Now added.

>    Should it say "see also" or "equivalent to"?

Well they aren't 100% equivalent due to template caching and sometimes the parameters only accept a type or have a stricter constraint than the __traits equivalents. (Using 'see also' doesn't imply they are the same).
